Resolve fenceposts for deployments together

0xmichalis opened the pull request:
In case of a deployment with zero maxSurge and a percent of maxUnavailable other than 100%, a deployment would block because it couldn't scale up (no surge) or scale down (maxUnavailable floored to zero). This PR resolves both fenceposts together, and in case of a zero maxSurge we ceil maxUnavailable, in case of a non-zero maxSurge, we continue flooring maxUnavailable.

@bgrant0607 @smarterclayton @janetkuo @mqliang PTAL

0xmichalis committed:
kubectl: set maxUnavailable to 1 if both fenceposts resolve to zero

Due to rounding down for maxUnavailable, we may end up with rolling updates
that have zero surge and unavailable pods something that 1) is not allowed
as per validation, 2) blocks updates. If we end up in such a situation
set maxUnavailable to 1 on the theory that surge might not work due to
quota.

kargakis commented:
cc: @ironcladlou

ironcladlou commented on pkg/util/deployment/deployment.go:
```diff
@@ -444,3 +444,35 @@ func WaitForObservedDeployment(getDeploymentFunc func() (*extensions.Deployment,
+	}
+	if surge == 0 {
+		// If a user has specified maxSurge to zero but maxUnavailable as a non-zero percent,
+		// then we want to allow scaling down and not block rollouts.
```
Not clear to me how this is being satisfied within the block- couldn't `unavailable` below be 0, causing the return of of surge=0/unavailable=0?

ironcladlou replied:
Nevermind, misunderstood how `GetValueFromIntOrPercent` works

bgrant0607 replied:
How about: 
* always round down for maxUnavailable
* if both are 0, then set unavailable to 1

Which would be more conservative.

Do we check in validation that both can't be 0/0%?

kargakis replied:
As a matter of fact, yes, we run this check during validation: https://github.com/kubernetes/kubernetes/blob/49f438bfac6174b00410739961c829eaa161b75c/pkg/apis/extensions/validation/validation.go#L269

bgrant0607 replied:
Doh! We should unify those functions with those in util/intstr/intstr.go.

So what about my other suggestions?

kargakis replied:
You mean switch back to use math.Ceil? Wouldn't this leave the window open for apps with both maxSurge and maxUnavailable specified being unavailable?

For example: 
desired=1, maxUnavailable=25%, maxSurge=25%, meaning we could scale down to zero while we could scale up using a surge pod. It seems better to disallow scaling down when we can scale up.

smarterclayton replied:
Regarding round down max unavailable (I might be misunderstanding your point).

We don't want to go to zero pods in a max surge 0 case unless max
unavailability is 100%.  Otherwise rolling deployments will take
downtime and aren't flexible to scaling.  We want a naive user with 2
pods to only take downtime if they explicitly ask for 100%
unavailability.  If scale = 1 we can round down when max surge is 0
(since setting scale to 1 with surge 0 means you know you will take
downtime).

bgrant0607 replied:
Sorry for not being clear.

The current code rounds maxUnavailable down and maxSurge up. I was proposing to leave them that way.

But if both unavailable pods and surge pods end up being zero, set unavailable to 1, on the theory that surge might not work due to quota.

smarterclayton replied:
Sorry, was reading on the API fields, not the code.

On Fri, Mar 4, 2016 at 12:37 PM, Brian Grant <notifications@github.com>
wrote:

> In pkg/util/deployment/deployment.go
> <https://github.com/kubernetes/kubernetes/pull/22516#discussion_r55069493>
> :
>
> > +// step. For example:
> > +//
> > +// 2 desired, max unavailable 1%, surge 0% - should scale old(-1), then new(+1), then old(-1), then new(+1)
> > +// 1 desired, max unavailable 1%, surge 0% - should scale old(-1), then new(+1)
> > +// 2 desired, max unavailable 25%, surge 1% - should scale new(+1), then old(-1), then new(+1), then old(-1)
> > +// 1 desired, max unavailable 25%, surge 1% - should scale new(+1), then old(-1)
> > +// 2 desired, max unavailable 0%, surge 1% - should scale new(+1), then old(-1), then new(+1), then old(-1)
> > +// 1 desired, max unavailable 0%, surge 1% - should scale new(+1), then old(-1)
> > +func ResolveFenceposts(maxSurge, maxUnavailable *intstrutil.IntOrString, desired int) (int, int, error) {
> > +	surge, err := intstrutil.GetValueFromIntOrPercent(maxSurge, desired, true)
> > +	if err != nil {
> > +		return 0, 0, err
> > +	}
> > +	if surge == 0 {
> > +		// If a user has specified maxSurge to zero but maxUnavailable as a non-zero percent,
> > +		// then we want to allow scaling down and not block rollouts.
>
> Sorry for not being clear.
>
> The current code rounds maxUnavailable down and maxSurge up. I was
> proposing to leave them that way.
>
> But if both unavailable pods and surge pods end up being zero, set
> unavailable to 1, on the theory that surge might not work due to quota.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/22516/files#r55069493>.
>

bgrant0607 replied:
I do not think we should round up maxUnavailable in general when surge is 0.

Ping @kargakis. I'd like to get this into 1.2 as a cherrypick.

smarterclayton replied:
Agree on not rounding up in general on maxUnavailable.  We just don't
want to go to available 0 ever when maxUnavailable != 100%

For some context here:

The scenario (which I expect is not uncommon) is our ingress L7
routers.  They are pods, running on host network, spread across a set
of nodes.  Those nodes have a keepalived and VVRP pool of virtual IPs
that get moved across nodes.  We can't surge (because there are only
so many hosts that qualify), but we don't want to set a numeric
maxUnavailable (because we only want to drop a few at a time, not all
of them).  We chose 25% because we expect in general that admins would
run sets of 3-4 routers in heavy load cases, and 2 routers otherwise.

In the future, we want to switch to daemon set for these, but would
need all the same guarantees.

kargakis replied:
maxUnavailable defaulting to one when everything has resolved to zero seems reasonable. I will update this tomorrow.

ironcladlou commented on pkg/kubectl/rolling_updater_test.go:
```diff
@@ -676,6 +676,27 @@ Scaling foo-v1 down to 0
 `,
 		},
+		{
+			name:        "2->2 25/1 maxSurge triumps maxUnavailable",
```
s/triumps/trumps

ironcladlou commented:
This seems like it'll work, couldn't hurt to get a second opinion

smarterclayton commented:
Test looks good to me, this I think should be part of 1.2 because 90% of our user's deployments will have some variation of 1 or 2 pods, and hit this fairly frequently.

bgrant0607 commented on pkg/util/deployment/deployment.go:
```diff
@@ -444,3 +444,35 @@ func WaitForObservedDeployment(getDeploymentFunc func() (*extensions.Deployment,
+// 1 desired, max unavailable 25%, surge 1% - should scale new(+1), then old(-1)
+// 2 desired, max unavailable 0%, surge 1% - should scale new(+1), then old(-1), then new(+1), then old(-1)
+// 1 desired, max unavailable 0%, surge 1% - should scale new(+1), then old(-1)
+func ResolveFenceposts(maxSurge, maxUnavailable *intstrutil.IntOrString, desired int) (int, int, error) {
```
Why pointers?

kargakis replied:
Already used in the rest of the IntOrString helpers so I opted not to change that here.

0xmichalis committed:
controller: resolve unavailable in conjuction with surge for deployments

Due to rounding down for maxUnavailable, we may end up with deployments
that have zero surge and unavailable pods something that 1) is not allowed
as per validation, 2) blocks deployments. If we end up in such a situation
set maxUnavailable to 1 on the theory that surge might not work due to
quota.

kargakis commented:
@bgrant0607 @smarterclayton updated to set maxUnavailable to one in case it resolves to zero due to rounding down. PTAL

bgrant0607 commented:
LGTM, thanks.

bgrant0607 commented:
I labeled this with cherrypick-candidate, so we don't forget to cherrypick this. We're not cherrypicking into the 1.2 branch quite yet.
